## Proyecto de aula
## Entrega 2: Ingenieria de caracteristicas

En este notebook se desarrollara la entrega 2 de la materia de analisis de datos de ingenieria de sistemas.

# Carga y comprension de datos




`a`. Carguen el archivo de datos en pandas

In [ ]:
import pandas as pd
from pathlib import Path 

url = "https://archive.ics.uci.edu/ml/machine-learning-databases/heart-disease/processed.cleveland.data"

columns = [
    "age", "sex", "cp", "trestbps", "chol", "fbs", "restecg",
    "thalach", "exang", "oldpeak", "slope", "ca", "thal", "target"
]


ROOT_DIR = Path().resolve().parent

DATA_DIR =  ROOT_DIR / "data/raw"

file_path = 'processed.cleveland.data'

df = pd.read_csv(
  
  url, names = columns,
  #na_values = ["?", "NA", "NaN", "nan", ""]
  
  )

df.head()

,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,target
0,63.0,1.0,1.0,145.0,233.0,1.0,2.0,150.0,0.0,2.3,3.0,0.0,6.0,0
1,67.0,1.0,4.0,160.0,286.0,0.0,2.0,108.0,1.0,1.5,2.0,3.0,3.0,2
2,67.0,1.0,4.0,120.0,229.0,0.0,2.0,129.0,1.0,2.6,2.0,2.0,7.0,1
3,37.0,1.0,3.0,130.0,250.0,0.0,0.0,187.0,0.0,3.5,3.0,0.0,3.0,0
4,41.0,0.0,2.0,130.0,204.0,0.0,2.0,172.0,0.0,1.4,1.0,0.0,3.0,0


`b`. Indique cúantas variables y registros tiene el dataframe, y asegúrense de que el tipo de dato de cada variable sea el esperado.


In [13]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       303 non-null    float64
 1   sex       303 non-null    float64
 2   cp        303 non-null    float64
 3   trestbps  303 non-null    float64
 4   chol      303 non-null    float64
 5   fbs       303 non-null    float64
 6   restecg   303 non-null    float64
 7   thalach   303 non-null    float64
 8   exang     303 non-null    float64
 9   oldpeak   303 non-null    float64
 10  slope     303 non-null    float64
 11  ca        303 non-null    str    
 12  thal      303 non-null    str    
 13  target    303 non-null    int64  
dtypes: float64(11), int64(1), str(2)
memory usage: 33.3 KB


`conclusion`: Se puede apreciar que en el df hay 14 columnas, y 303 filas.

`conclusion`: Se puede apreciar como las variables 'ca' y 'thal' tienen como tipo str, lo que es incorreco ya que son datos float64.
Debido a esto, se revisa y encuentra que estas variables presentan varios registros nulos, por lo que se definen como nulos con el objetivo de que pandas escoja correctamente el tipado.

In [29]:
df = pd.read_csv(
  
  url, names = columns,
  na_values = ["?", "NA", "NaN", "nan", ""]
  
  )

df.info()

<class 'pandas.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       303 non-null    float64
 1   sex       303 non-null    float64
 2   cp        303 non-null    float64
 3   trestbps  303 non-null    float64
 4   chol      303 non-null    float64
 5   fbs       303 non-null    float64
 6   restecg   303 non-null    float64
 7   thalach   303 non-null    float64
 8   exang     303 non-null    float64
 9   oldpeak   303 non-null    float64
 10  slope     303 non-null    float64
 11  ca        299 non-null    float64
 12  thal      301 non-null    float64
 13  target    303 non-null    int64  
dtypes: float64(13), int64(1)
memory usage: 33.3 KB


`c`. Reporte las estadisticas descriptivas de las variables numericas y categoricas

In [30]:
df.describe()

,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,target
count,303.000000,303.000000,303.000000,303.000000,303.000000,303.000000,303.000000,303.000000,303.000000,303.000000,303.000000,299.000000,301.000000,303.000000
mean,54.438944,0.679868,3.158416,131.689769,246.693069,0.148515,0.990099,149.607261,0.326733,1.039604,1.600660,0.672241,4.734219,0.937294
std,9.038662,0.467299,0.960126,17.599748,51.776918,0.356198,0.994971,22.875003,0.469794,1.161075,0.616226,0.937438,1.939706,1.228536
min,29.000000,0.000000,1.000000,94.000000,126.000000,0.000000,0.000000,71.000000,0.000000,0.000000,1.000000,0.000000,3.000000,0.000000
25%,48.000000,0.000000,3.000000,120.000000,211.000000,0.000000,0.000000,133.500000,0.000000,0.000000,1.000000,0.000000,3.000000,0.000000
50%,56.000000,1.000000,3.000000,130.000000,241.000000,0.000000,1.000000,153.000000,0.000000,0.800000,2.000000,0.000000,3.000000,0.000000
75%,61.000000,1.000000,4.000000,140.000000,275.000000,0.000000,2.000000,166.000000,1.000000,1.600000,2.000000,1.000000,7.000000,2.000000
max,77.000000,1.000000,4.000000,200.000000,564.000000,1.000000,2.000000,202.000000,1.000000,6.200000,3.000000,3.000000,7.000000,4.000000


`Numericas`: Se aprecia con df.info como cada variable presenta las estadisticas descriptivas anteriormente visualizadas.

`Categoricas`: Debido a que en el dataframe no hay variables categoricas, no se puede apreciar las estadisticas del mismo.

`d`. Identifique si el dataframe tiene datos duplicados

In [31]:
df.duplicated().sum()

np.int64(0)

`conclusion`: Se puede apreciar como el dataframe no tiene valores duplicados.

`e`. Identifique si hay variables con valores nulos. 

In [32]:
df.isnull().sum()

age         0
sex         0
cp          0
trestbps    0
chol        0
fbs         0
restecg     0
thalach     0
exang       0
oldpeak     0
slope       0
ca          4
thal        2
target      0
dtype: int64

`conclusion`: Se puede apreciar como las variables 'ca' y 'thal' presentan valores nulos, lo que causo aque al inicio se les asignara erroneamente el tipo de dato.

`f`. Reporten hallazgos en una celda de texto (Desarrollado a lo largo del taller) 

`g`. Descarten las variables que consideren irrelevantes para su modelo.

In [33]:
df.drop(columns=['ca', 'thal', 'fbs', 'age'], inplace=True)
df.head()

,sex,cp,trestbps,chol,restecg,thalach,exang,oldpeak,slope,target
0,1.0,1.0,145.0,233.0,2.0,150.0,0.0,2.3,3.0,0
1,1.0,4.0,160.0,286.0,2.0,108.0,1.0,1.5,2.0,2
2,1.0,4.0,120.0,229.0,2.0,129.0,1.0,2.6,2.0,1
3,1.0,3.0,130.0,250.0,0.0,187.0,0.0,3.5,3.0,0
4,0.0,2.0,130.0,204.0,2.0,172.0,0.0,1.4,1.0,0


`Conclusion`: Se descartaron las variables 'ca', 'thal y fbs, ya que las dos primeras presentan muchos datos nulos, por lo que se descarta, y se descarta fbs debido a que tiene poca varianza y mas del 85% es 0

`h`: Gestionen los datos duplicados y los datos nulos.

In [34]:
df = df.drop_duplicates()
df.isnull().sum()

sex         0
cp          0
trestbps    0
chol        0
restecg     0
thalach     0
exang       0
oldpeak     0
slope       0
target      0
dtype: int64

`Conclusion`: Ya que anteriormente se descartaron las unicas columnas con datos nulos, no se debe descartar ninguna columna en este paso. Ademas, se descartan los registros duplicados.

# 2. Analisis exploratorio de variables categóricas

`Conclusion`: Debido a que el dataframe no posee variables categoricas, el analisis exploratorio de variables categóricas no se puede realizar.

# 3	Análisis exploratorio de variables numéricas:

`a`. Identifiquen variables que sean (aproximadamente) normales y otras que tengan una distribución diferente.